### ISSUE TITLE
getting non-square matrices in error

### ISSUE THREAD

**Issue by khelfrich** [chebfun/chebfun#1575]
Hi,

I'm trying to solve a generalized eigenvalue problem Av = eBv where v = [w; b] and both w and b are functions. The script:

``` matlab
dom = [0 1];
z0   = 0.5;
lamu = 30;
lamb = 30;
Rimin = 0.15;
DU = sqrt(2*lamb/lamu^2/Rimin);
k = lamu*0.5;
Bz = chebfun(@(x)  0.5*lamb*sech(lamb*(x-z0)).^2, dom);
U  = chebfun(@(x)  DU/2*tanh(lamu*(x-z0)), dom);
Uzz= chebfun(@(x) -DU*lamu^2*tanh(lamu*(x-z0)).*sech(lamu*(x-z0)).^2, dom); 
A = chebop( @(z,w,b) [ -1i.*k.*U(z).*(diff(w,2)-k^2.*w)+1i.*k.*Uzz(z).*w - k^2.*b; ...
                        0*diff(b,2)-Bz(z).*w-1i*k.*U(z).*b], dom);
B = chebop( @(x,w,b) [ diff(w,2)-k^2.*w ; b], dom);
A.lbc = @(w,b)[ w; b];
A.rbc = @(w,b)[ w; b];
[V,e] = eigs(A,B,1,'LR');
```

returns the error  

```
Error using eig
For generalized eigenproblem EIG(A,B), A and B must be square.

Error in linop/eigs>getEigenvalues (line 351)
        [V, D] = eig(full(PA), full(PB));

Error in linop/eigs (line 233)
    [V, D, P] = getEigenvalues(discA, discB, k, sigma);

Error in chebop/eigs (line 109)
[varargout{1:nargout}] = eigs(L, varargin{:});
```

indicating that A and B are not square. They should be. Note the term 0*diff(b,2) in the definition for matrix A. This is done to make sure that b is interpreted as a function and not a scalar (rather then hack in chebop/linearize.m as suggested in https://github.com/chebfun/chebfun/issues/1566 that if used gives the same error as above).

FYI, this eigenvalue problem is a version the Taylor-Goldstein equation that is linear, rather than quadratic, in the eigenvalue. 

Any suggestions of where I have gone wrong would be helpful.

Karl

**Comment by asgeirbirkis** [comment]
@nickhale @tobydriscoll would you have expected this to work, i.e. should the projection algorithm for linear operators be able to deal with this?

In line 368 of `@linop/eigs`, which is the culprit for crashing and reads
```
[V, D] = eig(full(PA), full(PB));
```
I have just before the crash
```
K>> size(PA)
ans =
    68    66
K>> size(PB)
ans =
    68    66
```
Also, at that time, we have
```
K>> discA.dimAdjust
ans =
     2     0
     2     0
K>> discB.dimAdjust
ans =
     2     0
     2     0
```
and
```
K>> discA.source.diffOrder
ans =
     2     0
     0     0
K>> discB.source.diffOrder
ans =
     2     0
     0     0
```

A simpler version of the code that I could get to crash with the same error is:
```
dom = [0 1];
A = chebop( @(z,w,b) [ -1i.*(diff(w,2)-w)+1i.*w - b; ...
                        0*diff(b,2)-w-b], dom);
B = chebop( @(x,w,b) [ diff(w,2)-w ; b], dom);
A.lbc = @(w,b)[ w; b];
A.rbc = @(w,b)[ w; b];
[V,e] = eigs(A,B,1,'LR');
```

**Comment by khelfrich** [comment]
I'm a bit confused by your first question. If you are asking if the problem is correctly posed, the answer is yes. I have solved the same equations using standard finite-difference matrix methods in matlab.   

I also tried your slimmed down version of the problem and got the same error again, as did you. It seems to me that the problem has to do with the incorporation of the boundary conditions, that according to my slowly growing understanding of chebfun, should result in square matrices. But this is deep enough in chebfun that I can't easily see where the problem might be.

**Comment by chebfunuser** [comment]
@asgeirbirkis @khelfrich Has this problem been resolved? I am working on a similar problem and finding the same error.

**Comment by chebfunuser** [comment]
For future readers: I believe this problem has to do with the fact that @khelfrich is imposing two boundary conditions on b even though no derivatives in b exist in this problem. I do not think the introduction of the 0.*diff(b,2) term solves this problem. It probably introduces NaNs which the eigs function is attempting to eliminate or ignore in the projection. In other words, even though this problem may be well-posed and compatible with finite-difference methods, the spectral methods used in Chebfun lead to issues. The "N minus 2" rank phenomenon is common in incompressible spectral CFD formulations, particularly when pressure is involved. I suspect that is the issue here.

I have not yet resolved my issue, but it is very closely related to the above, and I will attempt to post an informative solution when it is resolved.

Bottom line: I believe the problem here is not with the algorithm, but rather with well-posedness. However, I believe Chebfun could be a bit more informative and helpful when it encounters this error.

**Comment by JamesMathews** [comment]
I too am having a similar problem, it still seems the example by @asgeirbirkis does not work in chebfun.

Do we have to just accept this kind of problem can't be solved in Chebfun?

**Comment by nickhale** [comment]
@chebfunuser is correct: the `0.*diff(b,2)` term does not fool `chebop` in to thinking the problem is second-order in `b` and hence giving a square system. However, you **can** fool it by using `realmin*diff(b, 2)`, i.e., 
```
dom = [0, 1];
A = chebop( @(z,w,b) [ -1i.*(diff(w,2)-w) + 1i.*w - b; ...
                        realmin*diff(b,2) - w - b], dom);
B = chebop( @(x,w,b) [ diff(w,2) - w ; b], dom);
A.lbc = @(w,b)[w ; b];
A.rbc = @(w,b)[w ; b];
[V, D] = eigs(A, B, 1);
% plot(real(V), '-', imag(V), '--'), shg
norm(A*V- B*V*D, inf)
```
which gives:
```
ans =
   4.9287e-09
```
![image](https://user-images.githubusercontent.com/2035010/31653147-d2792f94-b321-11e7-905d-12ff3533b484.png)
